test/concat_test.dart

```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
-      var second = new StreamController()
-        ..onCancel = () {
-          secondStreamClosed = true;
-        };
-      var subscription =
-          first.stream.transform(concat(second.stream)).listen((_) {});
-      await first.close();
-      await new Future(() {});
-      await subscription.cancel();
-      expect(secondStreamClosed, true);
-    });
-  });
+  var streamTypes = {
+    'single subscription': () => new StreamController(),
+    'broadcast': () => new StreamController.broadcast()
+  };
+  for (var firstType in streamTypes.keys) {
+    for (var secondType in streamTypes.keys) {
+      group('concat [$firstType] with [$secondType]', () {
+        StreamController first;
+        StreamController second;
+
+        List emittedValues;
+        bool firstCanceled;
+        bool secondCanceled;
+        bool secondListened;
+        bool isDone;
+        List errors;
+        Stream transformed;
+        StreamSubscription subscription;
+
+        setUp(() async {
+          firstCanceled = false;
+          secondCanceled = false;
+          secondListened = false;
+          first = streamTypes[firstType]()
+            ..onCancel = () {
+              firstCanceled = true;
+            };
+          second = streamTypes[secondType]()
+            ..onCancel = () {
+              secondCanceled = true;
+            }
+            ..onListen = () {
+              secondListened = true;
+            };
+          emittedValues = [];
+          errors = [];
+          isDone = false;
+          transformed = first.stream.transform(concat(second.stream));
+          subscription = transformed
+              .listen(emittedValues.add, onError: errors.add, onDone: () {
+            isDone = true;
+          });
+        });
+
+        test('adds all values from both streams', () async {
+          first..add(1)..add(2);
+          await first.close();
+          await new Future(() {});
+          second..add(3)..add(4);
+          await new Future(() {});
+          expect(emittedValues, [1, 2, 3, 4]);
+        });
+
+        test('Does not listen to second stream before first stream finishes',
+            () async {
+          expect(secondListened, false);
+          await first.close();
+          expect(secondListened, true);
+        });
+
+        test('cancels single-subscription first stream on cancel', () async {
+          await subscription.cancel();
+          expect(firstCanceled, true);
+        });
+
+        if (firstType == 'single-subscription') {
+          test('cancels single-subscription second stream on cancel', () async {
+            await first.close();
+            await subscription.cancel();
+            expect(secondCanceled, true);
+          });
+        }
+
+        test('closes stream after both inputs close', () async {
+          await first.close();
+          await second.close();
+          expect(isDone, true);
+        });
+
+        if (firstType == 'broadcast') {
```

Comment:
The current group names both types of streams so I don't think it would help the output. IMO the group would add a level of indent without much benefit.